### DISCUSSION TITLE
Adding search results to search registry for `n`/`N`

### DISCUSSION THREAD

**Discussion by lcpichette** [ibhagwan/fzf-lua#1632] [Q&A] [answered]
Hello!

I was wondering if it's possible to configure fzf-lua s.t. live_grep search results are added to the search register for easy navigation with `n` and `N`? I know I could do this with #442 in cycling files, but I just want the current buffer (grep_curbuf). 

The following is my attempt to translate cycling files into cycling results, but it just tells me that true isn't a valid option.

```lua
local map = vim.keymap.set
-- ... other mappigns
map("n", "/", function()
  require("fzf-lua").grep_curbuf({ fzf_opts = { ["--cycle"] = "true" } })
end)
```

Thank you to anyone reading this!

**Comment by ibhagwan** [comment]
I’m not sure I understand what you’re after, I think you mean search history but I don’t get how this is related to cycle?
https://github.com/ibhagwan/fzf-lua/wiki#how-do-i-setup-input-history-keybinds

Is this what you want?

**Reply by ibhagwan** [reply]
I think I understand what you want for that you’d need a custom function that sends the query to vim.fn.search, this will only work with live grep as it’s a proper regex, sending the fuzzy match query won’t mean much, something like the below.

Now this may or may not work as I quickly drafted this on an iPad, you main need to use `vim.fn.search` instead, but this started should get you where you want to be:
```lua
local fzf = require 'fzf-lua’
local utils = fzf.utils
fzf.lgrep_curbuf( {
  actions = { ['enter'] = function(sel, opts)
    local regex = utils.regex_to_magic(opts.last_search)
    local lnum = vim.api.nvim_win_get_cursor(0)[1]
    local line = vim.api.nvim_get_current_line(0)
    local reg = vim.regex(regex)
    reg:match_line(0, lnum, #line)
    -- also try:
    -- vim.api.nvim_win_set_cursor(0, { lnum-1, 0 })
    -- vim.fn.search(regex, "W")
  end }
})
```

**Reply by lcpichette** [reply]
I'm pretty new to neovim and lua, so I appreciate any patience with my poorly-written/understood lua in advance.

I use LazyVim, so in my init.lua I have `require("config.mappings")`. My `~/.config/nvim/lua/config/mappings.lua` file contains:
```lua
local custom_search = require("custom_search")
vim.keymap.set("n", "/", custom_search.searchFile, { desc = "Custom FZF lgrep logic" })
```

I have created a new file `~/.config/nvim/lua/custom_search.lua`:
```lua
local M = {}

local fzf = require("fzf-lua")
local utils = fzf.utils

function M.searchFile()
  fzf.lgrep_curbuf({
    actions = {
      ["enter"] = function(sel, opts)
        local regex = utils.regex_to_magic(opts.last_search)
        local lnum = vim.api.nvim_win_get_cursor(0)[1]
        local line = vim.api.nvim_get_current_line()
        local reg = vim.regex(regex)
        reg:match_line(0, lnum, #line)
        -- also try:
        -- vim.api.nvim_win_set_cursor(0, { lnum-1, 0 })
        -- vim.fn.search(regex, "W")
      end,
    },
  })
end

return M
```

Here's what it looks like to run `searchFile` (your logic) in the `custom_search` module (`/` being what I have it mapped to)

https://github.com/user-attachments/assets/6058a18c-6072-4ea0-9f1e-9be11b65ed6d

With the error:
```
[Fzf-lua] fn_selected threw an error: ...tte/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/utils.lua:215: attempt to index local 'str' (a nil value)│
    4             │stack traceback:                                                                                                                                 │
    5 function M.s│        ...ette/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:242: in function '__index'                                                   │
    6 ⎸ fzf.lgrep_│        ...tte/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/utils.lua:215: in function 'regex_to_magic'                                            │
    7 ⎸ ⎸ actions │        /Users/lucaspichette/.config/nvim/lua/custom_search.lua:10: in function 'action'                                                         │
H   8 ⎸ ⎸ ⎸ ["ente│        ...e/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/actions.lua:117: in function 'act'                                                       │
    9 ⎸ ⎸ ⎸ ⎸ loca│        ...ette/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:148: in function 'fn_selected'                                               │
   10 ⎸ ⎸ ⎸ ⎸ loca│        ...ette/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:235: in function <...ette/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua│
   11 ⎸ ⎸ ⎸ ⎸ loca│        [C]: in function 'xpcall'                                                                                                                │
   12 ⎸ ⎸ ⎸ ⎸ loca│        ...ette/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:234: in function <...ette/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua
```

If this is an issue with me just not understanding Lua let me know and I'll continue to work on it. But figured I'd atleast send this update in case maybe I am using your provided logic correctly, and there just needs to be a tweak to the logic.

I really appreciate how thorough and quick your response was btw!!  : )

**Reply by ibhagwan** [reply, marked as answer]
> If this is an issue with me just not understanding Lua let me know and I'll continue to work on it. But figured I'd atleast send this update in case maybe I am using your provided logic correctly, and there just needs to be a tweak to the logic.

It's lua and neovim and me just assuming what would work and typing it from the phone :-)

Tested and this works, can't guarantee edge cases you might need to debug this with special chars (although the magic pattern converter should do the trick), there probably is also a better way than manually feeding `/regex<CR>N` as keys but `vim.fn.search` or `vim.regex:match_line` didn't seem to do the trick, I also  tried `vim.fn.histadd("search", regex)`  - didn't work.
```lua
local fzf = require("fzf-lua")
local utils = fzf.utils
local actions = fzf.actions

fzf.lgrep_curbuf({
	actions = {
		["enter"] = function(sel, opts)
			actions.file_edit(sel, opts)
			if not opts.last_query or #opts.last_query == 0 then
				return
			end
			local regex = utils.regex_to_magic(opts.last_query)
			vim.api.nvim_feedkeys(
                          -- send `N` once after the search to go back to the selected line
                          vim.api.nvim_replace_termcodes("/" .. regex .. "<CR>N", true, false, true),
                       "n", true)
		end,
	},
})

```

**Reply by ibhagwan** [reply]
Couldn't find a better soltution than `feedkeys` for now but it seems to work nice, entered the regex `api.*\(`:
![image](https://github.com/user-attachments/assets/3bd19b7c-ab2b-4458-987e-f63689c56c85)

After pressing `enter` you get the correct regex highlighted (grep background is the next `n` match):
![image](https://github.com/user-attachments/assets/ca129277-6923-4e17-841e-382f5d7ff0a3)

**Reply by lcpichette** [reply]
Ah, I see. I did some more digging and I found the way to set the search registry `https://neovim.io/doc/user/builtin.html#setreg()`. I tried `/` hoping it would be a registry and it seems to work? Could have odd edge cases or unexpected side-effects, as I'm pretty new to this.

```lua
local M = {}

local fzf = require("fzf-lua")
local utils = fzf.utils
local actions = fzf.actions

function M.searchFile()
  fzf.lgrep_curbuf({
    actions = {
      ["enter"] = function(sel, opts)
        actions.file_edit(sel, opts)

        local last_query = opts.last_query
        if not last_query or #last_query == 0 then
          return
        end

        local regex = utils.regex_to_magic(last_query)

        vim.fn.setreg("/", regex)
        vim.api.nvim_feedkeys("N", "n", true)
        vim.api.nvim_feedkeys("n", "n", true)
      end,
    },
  })
end

return M
```

Both solutions work, but from an educational point of view I'm wondering what the difference between the two approaches is? Mine redundantly feeds two actions neovim (otherwise I find it doesn't actually work), but setting the `/` register explicitly like that makes sense to me. I'm not sure where your solution does that (but it clearly does).

Regardless, thank you a ton!! This is exactly the kind of functionality I was looking for. I really appreciate you going out of your way to help

**Reply by ibhagwan** [reply]
> Both solutions work, but from an educational point of view I'm wondering what the difference between the two approaches is? Mine redundantly feeds two actions neovim (otherwise I find it doesn't actually work), but setting the / register explicitly like that makes sense to me. I'm not sure where your solution does that (but it clearly does).

TBH, I’m not sure there’s a big difference, obviously feeding keys as if you just searched with `/` is going to work the same way but I think setreg is equally good, whatever works :)

> Regardless, thank you a ton!! This is exactly the kind of functionality I was looking for. I really appreciate you going out of your way to help

You’re welcome and and Happy Holidays.
